apps/emqx/src/emqx_channel.erl

```diff
@@ -2051,30 +2022,25 @@ maybe_resume_session(#channel{resuming = false}) ->
     ignore;
 maybe_resume_session(#channel{
     session = Session,
-    resuming = true,
-    pendings = Pendings,
+    resuming = ReplayContext,
     clientinfo = ClientInfo
 }) ->
-    {ok, Publishes, Session1} = emqx_session:replay(ClientInfo, Session),
-    case emqx_session:deliver(ClientInfo, Pendings, Session1) of
-        {ok, Session2} ->
-            {ok, Publishes, Session2};
-        {ok, More, Session2} ->
-            {ok, lists:append(Publishes, More), Session2}
-    end.
+    emqx_session:replay(ClientInfo, ReplayContext, Session).
```

Comment:
Discussion point. @thalesmg @ieQu1 

This is `ReplayContext` at play here. All this logic, including some [side-effects](https://github.com/emqx/emqx/pull/11592/commits/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc#diff-27c0844d0e2116cd96569c34369a4b485d68529f0d21cb075c19f8bc5a59f477L595) from `process_connect/2` was [moved to `emqx_session_mem:replay/3`](https://github.com/emqx/emqx/blob/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc/apps/emqx/src/emqx_session_mem.erl#L677-L690). Again, this needs a careful look that logic is kept (mostly) intact.

Reply:
The reference about the side-effects unfortunately was lost after some rebase, but I assume it's  `emqx_utils:drain_deliver()` and possibly `maybe_{,n}ack` in `session_mem`?

Reply:
Which one? One is still here, where `{takeover, 'end'}` is being handled. Another wasn't marked with such comment before for some reason:
https://github.com/emqx/emqx/blob/e089fda260574cac279faae6efaee798f87703be/apps/emqx/src/emqx_channel.erl#L603-L605

Reply:
I meant this link from the first message: https://github.com/emqx/emqx/pull/11592/commits/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc#diff-27c0844d0e2116cd96569c34369a4b485d68529f0d21cb075c19f8bc5a59f477L595

Reply:
Ah, sorry. 😅 Yes, that's the `emqx_utils:drain_deliver()` call, highlighted in my previous comment.